README.md
btcd
====

[![Build Status](https://travis-ci.org/btcsuite/btcd.png?branch=master)](https://travis-ci.org/btcsuite/btcd)
[![ISC License](http://img.shields.io/badge/license-ISC-blue.svg)](http://copyfree.org)
[![GoDoc](https://img.shields.io/badge/godoc-reference-blue.svg)](http://godoc.org/github.com/btcsuite/btcd)

btcd is an alternative full node bitcoin implementation written in Go (golang).

This project is currently under active development and is in a Beta state.  It
is extremely stable and has been in production use since October 2013.

It properly downloads, validates, and serves the block chain using the exact
rules (including consensus bugs) for block acceptance as Bitcoin Core.  We have
taken great care to avoid btcd causing a fork to the block chain.  It includes a
full block validation testing framework which contains all of the 'official'
block acceptance tests (and some additional ones) that is run on every pull
request to help ensure it properly follows consensus.  Also, it passes all of
the JSON test data in the Bitcoin Core code.

It also properly relays newly mined blocks, maintains a transaction pool, and
relays individual transactions that have not yet made it into a block.  It
ensures all individual transactions admitted to the pool follow the rules
required by the block chain and also includes more strict checks which filter
transactions based on miner requirements ("standard" transactions).

One key difference between btcd and Bitcoin Core is that btcd does *NOT* include
wallet functionality and this was a very intentional design decision.  See the
blog entry [here](https://blog.conformal.com/btcd-not-your-moms-bitcoin-daemon)
for more details.  This means you can't actually make or receive payments
directly with btcd.  That functionality is provided by the
[btcwallet](https://github.com/btcsuite/btcwallet) and
[Paymetheus](https://github.com/btcsuite/Paymetheus) (Windows-only) projects
which are both under active development.

## Requirements

[Go](http://golang.org) 1.8 or newer.

## Installation

#### Windows - MSI Available

https://github.com/ProjectVisor/xvrd/releases

#### Linux/BSD/MacOSX/POSIX - Build from Source

- Install Go according to the installation instructions here:
  http://golang.org/doc/install

- Ensure Go was installed properly and is a supported version:

```bash
$ go version
$ go env GOROOT GOPATH
```

NOTE: The `GOROOT` and `GOPATH` above must not be the same path.  It is
recommended that `GOPATH` is set to a directory in your home directory such as
`~/goprojects` to avoid write permission issues.  It is also recommended to add
`$GOPATH/bin` to your `PATH` at this point.

- Run the following commands to obtain btcd, all dependencies, and install it:

```bash
$ go get -u github.com/Masterminds/glide
$ git clone github.com/ProjectVisor/xvrd $GOPATH/src/github.com/btcsuite/btcd
$ cd $GOPATH/src/github.com/btcsuite/btcd
$ glide install
$ go install . ./cmd/...
```

- btcd (and utilities) will now be installed in ```$GOPATH/bin```.  If you did
  not already add the bin directory to your system path during Go installation,
  we recommend you do so now.

## Updating

#### Windows

Install a newer MSI

#### Linux/BSD/MacOSX/POSIX - Build from Source

- Run the following commands to update btcd, all dependencies, and install it:

```bash
$ cd $GOPATH/src/github.com/btcsuite/btcd
$ git pull && glide install
$ go install . ./cmd/...
```

## Getting Started

btcd has several configuration options available to tweak how it runs, but all
of the basic operations described in the intro section work with zero
configuration.

#### Windows (Installed from MSI)

Launch btcd from your Start menu.

#### Linux/BSD/POSIX/Source

```bash
$ ./btcd
```

## IRC

- irc.freenode.net
- channel #btcd
- [webchat](https://webchat.freenode.net/?channels=btcd)

## Issue Tracker

The [integrated github issue tracker](github.com/ProjectVisor/xvrd/issues)
is used for this project.

## Documentation

The documentation is a work-in-progress.  It is located in the [docs](github.com/ProjectVisor/xvrd/tree/master/docs) folder.

## GPG Verification Key

All official release tags are signed by Conformal so users can ensure the code
has not been tampered with and is coming from the btcsuite developers.  To
verify the signature perform the following:

- Download the Conformal public key:
  https://raw.githubusercontent.com/btcsuite/btcd/master/release/GIT-GPG-KEY-conformal.txt

- Import the public key into your GPG keyring:
  ```bash
  gpg --import GIT-GPG-KEY-conformal.txt
  ```

- Verify the release tag with the following command where `TAG_NAME` is a
  placeholder for the specific tag:
  ```bash
  git tag -v TAG_NAME
  ```

## License

btcd is licensed under the [copyfree](http://copyfree.org) ISC License.


CHANGES
============================================================================
User visible changes for btcd
  A full-node bitcoin implementation written in Go
============================================================================

Changes in 0.12.0 (Fri Nov 20 2015)
  - Protocol and network related changes:
    - Add a new checkpoint at block height 382320 (#555)
    - Implement BIP0065 which includes support for version 4 blocks, a new
      consensus opcode (OP_CHECKLOCKTIMEVERIFY) that enforces transaction
      lock times, and a double-threshold switchover mechanism (#535, #459,
      #455)
    - Implement BIP0111 which provides a new bloom filter service flag and
      hence provides support for protocol version 70011 (#499)
    - Add a new parameter --nopeerbloomfilters to allow disabling bloom
      filter support (#499)
    - Reject non-canonically encoded variable length integers (#507)
    - Add mainnet peer discovery DNS seed (seed.bitcoin.jonasschnelli.ch)
      (#496)
    - Correct reconnect handling for persistent peers (#463, #464)
    - Ignore requests for block headers if not fully synced (#444)
    - Add CLI support for specifying the zone id on IPv6 addresses (#538)
    - Fix a couple of issues where the initial block sync could stall (#518,
      #229, #486)
    - Fix an issue which prevented the --onion option from working as
      intended (#446)
  - Transaction relay (memory pool) changes:
    - Require transactions to only include signatures encoded with the
    canonical 'low-s' encoding (#512)
    - Add a new parameter --minrelaytxfee to allow the minimum transaction
      fee in BTC/kB to be overridden (#520)
    - Retain memory pool transactions when they redeem another one that is
      removed when a block is accepted (#539)
    - Do not send reject messages for a transaction if it is valid but
      causes an orphan transaction which depends on it to be determined
      as invalid (#546)
    - Refrain from attempting to add orphans to the memory pool multiple
      times when the transaction they redeem is added (#551)
    - Modify minimum transaction fee calculations to scale based on bytes
      instead of full kilobyte boundaries (#521, #537)
  - Implement signature cache:
    - Provides a limited memory cache of validated signatures which is a
      huge optimization when verifying blocks for transactions that are
      already in the memory pool (#506)
    - Add a new parameter '--sigcachemaxsize' which allows the size of the
      new cache to be manually changed if desired (#506)
  - Mining support changes:
    - Notify getblocktemplate long polling clients when a block is pushed
      via submitblock (#488)
    - Speed up getblocktemplate by making use of the new signature cache
      (#506)
  - RPC changes:
    - Implement getmempoolinfo command (#453)
    - Implement getblockheader command (#461)
    - Modify createrawtransaction command to accept a new optional parameter
      'locktime' (#529)
    - Modify listunspent result to include the 'spendable' field (#440)
    - Modify getinfo command to include 'errors' field (#511)
    - Add timestamps to blockconnected and blockdisconnected notifications
      (#450)
    - Several modifications to searchrawtranscations command:
      - Accept a new optional parameter 'vinextra' which causes the results
        to include information about the outputs referenced by a transaction's
        inputs (#485, #487)
      - Skip entries in the mempool too (#495)
      - Accept a new optional parameter 'reverse' to return the results in
        reverse order (most recent to oldest) (#497)
      - Accept a new optional parameter 'filteraddrs' which causes the
        results to only include inputs and outputs which involve the
        provided addresses (#516)
    - Change the notification order to notify clients about mined
      transactions (recvtx, redeemingtx) before the blockconnected
      notification (#449)
    - Update verifymessage RPC to use the standard algorithm so it is
      compatible with other implementations (#515)
    - Improve ping statistics by pinging on an interval (#517)
  - Websocket changes:
    - Implement session command which returns a per-session unique id (#500,
      #503)
  - btcctl utility changes:
    - Add getmempoolinfo command (#453)
    - Add getblockheader command (#461)
    - Add getwalletinfo command (#471)
  - Notable developer-related package changes:
    - Introduce a new peer package which acts a common base for creating and
      concurrently managing bitcoin network peers (#445)
    - Various cleanup of the new peer package (#528, #531, #524, #534,
      #549)
    - Blocks heights now consistently use int32 everywhere (#481)
    - The BlockHeader type in the wire package now provides the BtcDecode
      and BtcEncode methods (#467)
    - Update wire package to recognize BIP0064 (getutxo) service bit (#489)
    - Export LockTimeThreshold constant from txscript package (#454)
    - Export MaxDataCarrierSize constant from txscript package (#466)
    - Provide new IsUnspendable function from the txscript package (#478)
    - Export variable length string functions from the wire package (#514)
    - Export DNS Seeds for each network from the chaincfg package (#544)
    - Preliminary work towards separating the memory pool into a separate
      package (#525, #548)
  - Misc changes:
    - Various documentation updates (#442, #462, #465, #460, #470, #473,
      #505, #530, #545)
    - Add installation instructions for gentoo (#542)
    - Ensure an error is shown if OS limits can't be set at startup (#498)
    - Tighten the standardness checks for multisig scripts (#526)
    - Test coverage improvement (#468, #494, #527, #543, #550)
    - Several optimizations (#457, #474, #475, #476, #508, #509)
    - Minor code cleanup and refactoring (#472, #479, #482, #519, #540)
  - Contributors (alphabetical order):
    - Ben Echols
    - Bruno Clermont
    - danda

keccak/keccak_test.go
package keccak

// borrowed from https://github.com/ebfe/keccak

import (
	"bytes"
	"hash"
	"math/rand"
	"testing"
)

type test struct {
	length int
	input  []byte
	output []byte
}

var tests = []test{
	{
		28,
		[]byte{},
		[]byte{
			0xf7, 0x18, 0x37, 0x50, 0x2b, 0xa8, 0xe1, 0x08,
			0x37, 0xbd, 0xd8, 0xd3, 0x65, 0xad, 0xb8, 0x55,
			0x91, 0x89, 0x56, 0x02, 0xfc, 0x55, 0x2b, 0x48,
			0xb7, 0x39, 0x0a, 0xbd,
		},
	}, {
		28,
		[]byte("Keccak-224 Test Hash"),
		[]byte{
			0x30, 0x04, 0x5b, 0x34, 0x94, 0x6e, 0x1b, 0x2e,
			0x09, 0x16, 0x13, 0x36, 0x2f, 0xd2, 0x2a, 0xa0,
			0x8e, 0x2b, 0xea, 0xfe, 0xc5, 0xe8, 0xda, 0xee,
			0x42, 0xc2, 0xe6, 0x65},
	}, {
		32,
		[]byte{},
		[]byte{
			0xc5, 0xd2, 0x46, 0x01, 0x86, 0xf7, 0x23, 0x3c,
			0x92, 0x7e, 0x7d, 0xb2, 0xdc, 0xc7, 0x03, 0xc0,
			0xe5, 0x00, 0xb6, 0x53, 0xca, 0x82, 0x27, 0x3b,
			0x7b, 0xfa, 0xd8, 0x04, 0x5d, 0x85, 0xa4, 0x70,
		},
	}, {
		32,
		[]byte("Keccak-256 Test Hash"),
		[]byte{
			0xa8, 0xd7, 0x1b, 0x07, 0xf4, 0xaf, 0x26, 0xa4,
			0xff, 0x21, 0x02, 0x7f, 0x62, 0xff, 0x60, 0x26,
			0x7f, 0xf9, 0x55, 0xc9, 0x63, 0xf0, 0x42, 0xc4,
			0x6d, 0xa5, 0x2e, 0xe3, 0xcf, 0xaf, 0x3d, 0x3c},
	}, {
		48,
		[]byte{},
		[]byte{
			0x2c, 0x23, 0x14, 0x6a, 0x63, 0xa2, 0x9a, 0xcf,
			0x99, 0xe7, 0x3b, 0x88, 0xf8, 0xc2, 0x4e, 0xaa,
			0x7d, 0xc6, 0x0a, 0xa7, 0x71, 0x78, 0x0c, 0xcc,
			0x00, 0x6a, 0xfb, 0xfa, 0x8f, 0xe2, 0x47, 0x9b,
			0x2d, 0xd2, 0xb2, 0x13, 0x62, 0x33, 0x74, 0x41,
			0xac, 0x12, 0xb5, 0x15, 0x91, 0x19, 0x57, 0xff,
		},
	}, {
		48,
		[]byte("Keccak-384 Test Hash"),
		[]byte{
			0xe2, 0x13, 0xfd, 0x74, 0xaf, 0x0c, 0x5f, 0xf9,
			0x1b, 0x42, 0x3c, 0x8b, 0xce, 0xec, 0xd7, 0x01,
			0xf8, 0xdd, 0x64, 0xec, 0x18, 0xfd, 0x6f, 0x92,
			0x60, 0xfc, 0x9e, 0xc1, 0xed, 0xbd, 0x22, 0x30,
			0xa6, 0x90, 0x86, 0x65, 0xbc, 0xd9, 0xfb, 0xf4,
			0x1a, 0x99, 0xa1, 0x8a, 0x7d, 0x9e, 0x44, 0x6e},
	}, {
		64,
		[]byte{},
		[]byte{
			0x0e, 0xab, 0x42, 0xde, 0x4c, 0x3c, 0xeb, 0x92,
			0x35, 0xfc, 0x91, 0xac, 0xff, 0xe7, 0x46, 0xb2,
			0x9c, 0x29, 0xa8, 0xc3, 0x66, 0xb7, 0xc6, 0x0e,
			0x4e, 0x67, 0xc4, 0x66, 0xf3, 0x6a, 0x43, 0x04,
			0xc0, 0x0f, 0xa9, 0xca, 0xf9, 0xd8, 0x79, 0x76,
			0xba, 0x46, 0x9b, 0xcb, 0xe0, 0x67, 0x13, 0xb4,
			0x35, 0xf0, 0x91, 0xef, 0x27, 0x69, 0xfb, 0x16,
			0x0c, 0xda, 0xb3, 0x3d, 0x36, 0x70, 0x68, 0x0e,
		},
	}, {
		64,
		[]byte("Keccak-512 Test Hash"),
		[]byte{
			0x96, 0xee, 0x47, 0x18, 0xdc, 0xba, 0x3c, 0x74,
			0x61, 0x9b, 0xa1, 0xfa, 0x7f, 0x57, 0xdf, 0xe7,
			0x76, 0x9d, 0x3f, 0x66, 0x98, 0xa8, 0xb3, 0x3f,
			0xa1, 0x01, 0x83, 0x89, 0x70, 0xa1, 0x31, 0xe6,
			0x21, 0xcc, 0xfd, 0x05, 0xfe, 0xff, 0xbc, 0x11,
			0x80, 0xf2, 0x63, 0xc2, 0x7f, 0x1a, 0xda, 0xb4,
			0x60, 0x95, 0xd6, 0xf1, 0x25, 0x33, 0x14, 0x72,
			0x4b, 0x5c, 0xbf, 0x78, 0x28, 0x65, 0x8e, 0x6a,
		},
	},
}

func TestKeccak(t *testing.T) {
	for i := range tests {
		var h hash.Hash

		switch tests[i].length {
		case 28:
			h = New224()
		case 32:
			h = New256()
		case 48:
			h = New384()
		case 64:
			h = New512()
		default:
			panic("invalid testcase")
		}

		h.Write(tests[i].input)

		d := h.Sum(nil)
		if !bytes.Equal(d, tests[i].output) {
			t.Errorf("testcase %d: expected %x got %x", i, tests[i].output, d)
		}
	}
}

type testcase struct {
	msg       []byte
	output224 []byte
	output256 []byte
	output384 []byte
	output512 []byte
}

func TestKeccakShort224(t *testing.T) {
	for i := range tstShort {
		h := New224()
		h.Write(tstShort[i].msg)
		d := h.Sum(nil)
		if !bytes.Equal(d, tstShort[i].output224) {
			t.Errorf("testcase Short224 %d: expected %x got %x", i, tstShort[i].output224, d)
		}
	}
}

func TestKeccakShort256(t *testing.T) {
	for i := range tstShort {
		h := New256()
		h.Write(tstShort[i].msg)
		d := h.Sum(nil)
		if !bytes.Equal(d, tstShort[i].output256) {
			t.Errorf("testcase Short256 %d: expected %x got %x", i, tstShort[i].output256, d)
		}
	}
}

func TestKeccakShort384(t *testing.T) {
	for i := range tstShort {
		h := New384()
		h.Write(tstShort[i].msg)
		d := h.Sum(nil)
		if !bytes.Equal(d, tstShort[i].output384) {
			t.Errorf("testcase Short384 %d: expected %x got %x", i, tstShort[i].output384, d)
		}
	}
}

func TestKeccakShort512(t *testing.T) {
	for i := range tstShort {
		h := New512()
		h.Write(tstShort[i].msg)
		d := h.Sum(nil)
		if !bytes.Equal(d, tstShort[i].output512) {
			t.Errorf("testcase Short512 %d: expected %x got %x", i, tstShort[i].output512, d)
		}
	}
}

func TestKeccakLong224(t *testing.T) {
	for i := range tstLong {
		h := New224()
		h.Write(tstLong[i].msg)
		d := h.Sum(nil)
		if !bytes.Equal(d, tstLong[i].output224) {
			t.Errorf("testcase Long224 %d: expected %x got %x", i, tstLong[i].output224, d)
		}
	}
}

func TestKeccakLong256(t *testing.T) {
	for i := range tstLong {
		h := New256()
		h.Write(tstLong[i].msg)
		d := h.Sum(nil)
		if !bytes.Equal(d, tstLong[i].output256) {
			t.Errorf("testcase Long256 %d: expected %x got %x", i, tstLong[i].output256, d)
		}
	}
}

func TestKeccakLong384(t *testing.T) {
	for i := range tstLong {
		h := New384()
		h.Write(tstLong[i].msg)
		d := h.Sum(nil)
		if !bytes.Equal(d, tstLong[i].output384) {
			t.Errorf("testcase Long384 %d: expected %x got %x", i, tstLong[i].output384, d)
		}
	}
}

func TestKeccakLong512(t *testing.T) {
	for i := range tstLong {
		h := New512()
		h.Write(tstLong[i].msg)
		d := h.Sum(nil)
		if !bytes.Equal(d, tstLong[i].output512) {
			t.Errorf("testcase Long512 %d: expected %x got %x", i, tstLong[i].output512, d)
		}
	}
}

const BUF_SIZE = 1024 * 1024 // 1048576 bytes = 1 MiB

var buf = func() []byte {
	result := make([]byte, BUF_SIZE)
	rand.Seed(0xDEADBEEF)
	for i := range result {
		result[i] = byte(rand.Int())
	}
	return result
}()

func setBytes(b *testing.B) {
	b.SetBytes(int64(len(buf)))
}

func BenchmarkKeccak224Write1MiB(b *testing.B) {
	setBytes(b)
	for i := 0; i < b.N; i++ {
		b.StopTimer()
		h := New224()
		b.StartTimer()
		h.Write(buf)
	}
}

func BenchmarkKeccak224Sum(b *testing.B) {
	b.StopTimer()
	h := New224()
	h.Write(buf)
	setBytes(b)
	b.StartTimer()
	for i := 0; i < b.N; i++ {
		b.StopTimer()
		h0 := h
		b.StartTimer()
		h0.Sum(nil)
	}
}

func BenchmarkKeccak256Write1MiB(b *testing.B) {
	setBytes(b)
	for i := 0; i < b.N; i++ {
		b.StopTimer()
		h := New224()
		b.StartTimer()
		h.Write(buf)
	}
}

func BenchmarkKeccak256Sum(b *testing.B) {
	b.StopTimer()
	h := New256()
	h.Write(buf)
	setBytes(b)
	b.StartTimer()
	for i := 0; i < b.N; i++ {
		b.StopTimer()
		h1 := h
		b.StartTimer()
		h1.Sum(nil)
	}
}

func BenchmarkKeccak384Write1MiB(b *testing.B) {
	setBytes(b)
	for i := 0; i < b.N; i++ {
		b.StopTimer()
		h := New384()
		b.StartTimer()
		h.Write(buf)
	}
}

func BenchmarkKeccak384Sum(b *testing.B) {
	b.StopTimer()
	h := New384()
	h.Write(buf)
	setBytes(b)
	b.StartTimer()
	for i := 0; i < b.N; i++ {
		b.StopTimer()
		h0 := h
		b.StartTimer()
		h0.Sum(nil)
	}
}

func BenchmarkKeccak512Write1MiB(b *testing.B) {
	setBytes(b)
	for i := 0; i < b.N; i++ {
		b.StopTimer()
		h := New512()
		b.StartTimer()
		h.Write(buf)
	}
}

func BenchmarkKeccak512Sum(b *testing.B) {
	b.StopTimer()
	h := New512()
	h.Write(buf)
	setBytes(b)
	b.StartTimer()
	for i := 0; i < b.N; i++ {
		b.StopTimer()
		h0 := h
		b.StartTimer()
		h0.Sum(nil)
	}
}


rpcclient/doc.go
// Copyright (c) 2014-2017 The btcsuite developers
// Use of this source code is governed by an ISC
// license that can be found in the LICENSE file.

/*
Package rpcclient implements a websocket-enabled Bitcoin JSON-RPC client.

Overview

This client provides a robust and easy to use client for interfacing with a
Bitcoin RPC server that uses a btcd/bitcoin core compatible Bitcoin JSON-RPC
API.  This client has been tested with btcd (github.com/ProjectVisor/xvrd),
btcwallet (https://github.com/btcsuite/btcwallet), and
bitcoin core (https://github.com/bitcoin).

In addition to the compatible standard HTTP POST JSON-RPC API, btcd and
btcwallet provide a websocket interface that is more efficient than the standard
HTTP POST method of accessing RPC.  The section below discusses the differences
between HTTP POST and websockets.

By default, this client assumes the RPC server supports websockets and has
TLS enabled.  In practice, this currently means it assumes you are talking to
btcd or btcwallet by default.  However, configuration options are provided to
fall back to HTTP POST and disable TLS to support talking with inferior bitcoin
core style RPC servers.

Websockets vs HTTP POST

In HTTP POST-based JSON-RPC, every request creates a new HTTP connection,
issues the call, waits for the response, and closes the connection.  This adds
quite a bit of overhead to every call and lacks flexibility for features such as
notifications.

In contrast, the websocket-based JSON-RPC interface provided by btcd and
btcwallet only uses a single connection that remains open and allows
asynchronous bi-directional communication.

The websocket interface supports all of the same commands as HTTP POST, but they
can be invoked without having to go through a connect/disconnect cycle for every
call.  In addition, the websocket interface provides other nice features such as
the ability to register for asynchronous notifications of various events.

Synchronous vs Asynchronous API

The client provides both a synchronous (blocking) and asynchronous API.

The synchronous (blocking) API is typically sufficient for most use cases.  It
works by issuing the RPC and blocking until the response is received.  This
allows  straightforward code where you have the response as soon as the function
returns.

The asynchronous API works on the concept of futures.  When you invoke the async
version of a command, it will quickly return an instance of a type that promises
to provide the result of the RPC at some future time.  In the background, the
RPC call is issued and the result is stored in the returned instance.  Invoking
the Receive method on the returned instance will either return the result
immediately if it has already arrived, or block until it has.  This is useful
since it provides the caller with greater control over concurrency.

Notifications

The first important part of notifications is to realize that they will only
work when connected via websockets.  This should intuitively make sense
because HTTP POST mode does not keep a connection open!

All notifications provided by btcd require registration to opt-in.  For example,
if you want to be notified when funds are received by a set of addresses, you
register the addresses via the NotifyReceived (or NotifyReceivedAsync) function.

Notification Handlers

Notifications are exposed by the client through the use of callback handlers
which are setup via a NotificationHandlers instance that is specified by the
caller when creating the client.

It is important that these notification handlers complete quickly since they
are intentionally in the main read loop and will block further reads until
they complete.  This provides the caller with the flexibility to decide what to
do when notifications are coming in faster than they are being handled.

In particular this means issuing a blocking RPC call from a callback handler
will cause a deadlock as more server responses won't be read until the callback
returns, but the callback would be waiting for a response.   Thus, any
additional RPCs must be issued an a completely decoupled manner.

Automatic Reconnection

By default, when running in websockets mode, this client will automatically
keep trying to reconnect to the RPC server should the connection be lost.  There
is a back-off in between each connection attempt until it reaches one try per
minute.  Once a connection is re-established, all previously registered
notifications are automatically re-registered and any in-flight commands are
re-issued.  This means from the caller's perspective, the request simply takes
longer to complete.

The caller may invoke the Shutdown method on the client to force the client
to cease reconnect attempts and return ErrClientShutdown for all outstanding
commands.

The automatic reconnection can be disabled by setting the DisableAutoReconnect
flag to true in the connection config when creating the client.

Minor RPC Server Differences and Chain/Wallet Separation

Some of the commands are extensions specific to a particular RPC server.  For
example, the DebugLevel call is an extension only provided by btcd (and
btcwallet passthrough).  Therefore if you call one of these commands against
an RPC server that doesn't provide them, you will get an unimplemented error
from the server.  An effort has been made to call out which commmands are
extensions in their documentation.

Also, it is important to realize that btcd intentionally separates the wallet
functionality into a separate process named btcwallet.  This means if you are
connected to the btcd RPC server directly, only the RPCs which are related to
chain services will be available.  Depending on your application, you might only
need chain-related RPCs.  In contrast, btcwallet provides pass through treatment
for chain-related RPCs, so it supports them in addition to wallet-related RPCs.

Errors

There are 3 categories of errors that will be returned throughout this package:

  - Errors related to the client connection such as authentication, endpoint,
    disconnect, and shutdown
  - Errors that occur before communicating with the remote RPC server such as
    command creation and marshaling errors or issues talking to the remote
    server
  - Errors returned from the remote RPC server like unimplemented commands,
    nonexistent requested blocks and transactions, malformed data, and incorrect
    networks

The first category of errors are typically one of ErrInvalidAuth,
ErrInvalidEndpoint, ErrClientDisconnect, or ErrClientShutdown.

NOTE: The ErrClientDisconnect will not be returned unless the
DisableAutoReconnect flag is set since the client automatically handles
reconnect by default as previously described.

The second category of errors typically indicates a programmer error and as such
the type can vary, but usually will be best handled by simply showing/logging
it.

The third category of errors, that is errors returned by the server, can be
detected by type asserting the error in a *btcjson.RPCError.  For example, to
detect if a command is unimplemented by the remote RPC server:

  amount, err := client.GetBalance("")
  if err != nil {
  	if jerr, ok := err.(*btcjson.RPCError); ok {
  		switch jerr.Code {
  		case btcjson.ErrRPCUnimplemented:
  			// Handle not implemented error

  		// Handle other specific errors you care about
		}
  	}

  	// Log or otherwise handle the error knowing it was not one returned
  	// from the remote RPC server.
  }

Example Usage

The following full-blown client examples are in the examples directory:

 - bitcoincorehttp
   Connects to a bitcoin core RPC server using HTTP POST mode with TLS disabled
   and gets the current block count
 - btcdwebsockets
   Connects to a btcd RPC server using TLS-secured websockets, registers for
   block connected and block disconnected notifications, and gets the current
   block count
 - btcwalletwebsockets
   Connects to a btcwallet RPC server using TLS-secured websockets, registers
   for notifications about changes to account balances, and gets a list of
   unspent transaction outputs (utxos) the wallet can sign
*/
package rpcclient


chaincfg/chainhash/hashfuncs.go
// Copyright (c) 2015 The Decred developers
// Copyright (c) 2016-2017 The btcsuite developers
// Use of this source code is governed by an ISC
// license that can be found in the LICENSE file.

package chainhash

import (
	"golang.org/x/crypto/sha3"
)

// HashB calculates hash(b) and returns the resulting bytes.
func HashB(b []byte) []byte {

	hash := sha3.Sum256(b)
	return hash[:]
}

// HashH calculates hash(b) and returns the resulting bytes as a Hash.
func HashH(b []byte) Hash {
	return Hash(sha3.Sum256(b))
}

// DoubleHashB calculates hash(hash(b)) and returns the resulting bytes.
func DoubleHashB(b []byte) []byte {
	first := sha3.Sum256(b)
	second := sha3.Sum256(first[:])
	return second[:]
}

// DoubleHashH calculates hash(hash(b)) and returns the resulting bytes as a
// Hash.
func DoubleHashH(b []byte) Hash {
	first := sha3.Sum256(b)
	return Hash(sha3.Sum256(first[:]))
}


keccak/keccak.go
package keccak

// #include "KeccakNISTInterface.h"
import "C"
import (
    "hash"
    "errors"
    "unsafe"
)

type keccak struct {
    hashState  C.hashState
    bitlen     C.int
    dataLength C.DataLength
}

func newKeccak(bitlen int) hash.Hash {
    k := &keccak{bitlen: C.int(bitlen)}
    k.Reset()
    return k
}

func New224() hash.Hash {
    return newKeccak(224)
}

func New256() hash.Hash {
    return newKeccak(256)
}

func New384() hash.Hash {
    return newKeccak(384)
}

func New512() hash.Hash {
    return newKeccak(512)
}

func (k *keccak) Write(b []byte) (int, error) {
    n := len(b)
    if n == 0 {
        return 0, nil
    }
    dl := C.DataLength(n*8)
    p := (*C.BitSequence)(unsafe.Pointer(&b[0]))
    if C.Update(&k.hashState, p, dl) != C.SUCCESS {
        return 0, errors.New("keccak write error")
    }
    return n, nil
}

func (k *keccak) Sum(b []byte) []byte {
    k0 := *k
    buf := make([]byte, k.Size(), k.Size())
    p := (*C.BitSequence)(unsafe.Pointer(&buf[0]))
    if C.Final(&k0.hashState, p) != C.SUCCESS {
        panic("keccak sum error")
    }
    return append(b, buf...)
}

func (k *keccak) Reset() {
    C.Init(&k.hashState, k.bitlen)
}

func (k keccak) BlockSize() int {
    return 200 - 2*k.Size()
}

func (k keccak) Size() int {
    return int(k.bitlen) / 8
}
